src/meta/src/stream/scale.rs

```diff
@@ -0,0 +1,827 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::min;
+use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
+
+use async_recursion::async_recursion;
+use itertools::Itertools;
+use risingwave_common::bail;
+use risingwave_common::catalog::TableId;
+use risingwave_common::types::{ParallelUnitId, VIRTUAL_NODE_COUNT};
+use risingwave_pb::common::{ActorInfo, ParallelUnit, ParallelUnitMapping, WorkerNode, WorkerType};
+use risingwave_pb::meta::table_fragments::fragment::FragmentDistributionType;
+use risingwave_pb::meta::table_fragments::{ActorState, ActorStatus, Fragment};
+use risingwave_pb::stream_plan::barrier::Mutation;
+use risingwave_pb::stream_plan::stream_node::NodeBody;
+use risingwave_pb::stream_plan::{
+    ActorMapping, DispatcherType, PauseMutation, ResumeMutation, StreamActor, StreamNode,
+};
+use risingwave_pb::stream_service::{
+    BroadcastActorInfoTableRequest, BuildActorsRequest, HangingChannel, UpdateActorsRequest,
+};
+use uuid::Uuid;
+
+use crate::barrier::{Command, Reschedule};
+use crate::manager::{IdCategory, WorkerId};
+use crate::model::{ActorId, DispatcherId, FragmentId, TableFragments};
+use crate::storage::MetaStore;
+use crate::stream::GlobalStreamManager;
+use crate::MetaResult;
+
+#[derive(Debug)]
+pub struct ParallelUnitReschedule {
+    pub added_parallel_units: Vec<ParallelUnitId>,
+    pub removed_parallel_units: Vec<ParallelUnitId>,
+}
+
+impl<S> GlobalStreamManager<S>
+where
+    S: MetaStore,
+{
+    pub async fn reschedule_actors(
+        &self,
+        reschedule: HashMap<FragmentId, ParallelUnitReschedule>,
+    ) -> MetaResult<()> {
+        // First we create the necessary indexes
+        let table_ids = self
+            .fragment_manager
+            .find_table_ids_for_fragment_ids(reschedule.keys().cloned().collect())
+            .await
+            .values()
+            .cloned()
+            .collect();
+
```

Comment:
Add assert for migration of source fragment and mview fragment with downstream fragment, they're not supported now.